flow/aggregates.py

```diff
@@ -0,0 +1,266 @@
+            aggregated_jobs = list(filter(self._aggregate._select, aggregated_jobs))
+        if self._aggregate._sort_by is not None:
+            aggregated_jobs = list(sorted(aggregated_jobs,
+                                          key=lambda job: job.sp[self._aggregate._sort_by],
+                                          reverse=bool(self._aggregate._reverse_order)))
+
+        aggregated_jobs = self._aggregate._aggregator([job for job in aggregated_jobs])
+        self._create_nested_aggregate_list(aggregated_jobs, project)
+
+    def _create_nested_aggregate_list(self, aggregated_jobs, project):
+        # This method converts the returned subset of jobs as an Iterable
+        # from an aggregator function to a subset of jobs as tuple.
+        aggregated_jobs = list(aggregated_jobs)
+        nested_aggregates = []
+
+        for aggregate in aggregated_jobs:
+            try:
+                filter_aggregate = []
+                for job in aggregate:
+                    if job is None:
+                        continue
+                    elif job not in project:
+                        raise LookupError(f'The signac job {str(job)} not found in {project}')
+                    filter_aggregate.append(job)
+                filter_aggregate = tuple(filter_aggregate)
+                nested_aggregates.append(filter_aggregate)
+            except TypeError:  # aggregate is not iterable
+                raise ValueError("Invalid aggregator function provided by "
+                                 "the user.")
+
+        self._aggregates = nested_aggregates
+
+    def __hash__(self):
+        blob = str(hash(self._aggregate))
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+
+class _StoreDefaultAggregate:
+    """This class holds the information of the project associated with
+    an operation function aggregated by the default aggregates i.e.
+    :py:class:`Aggregate.groupsof(1)`.
+
+    When iterated through it's instance, it yields a tuple of a single job from
+    the Project.
+    """
+    def __init__(self):
+        self._project = None
+
+    def __iter__(self):
+        for job in self._project:
+            yield (job,)
+
+    def __call__(self, jobs):
+        # We have to store self._project is a instance of this class
+        # is called. This is because we will then iterate over that
+        # project in order to return an aggregates of one.
+        if isinstance(jobs, list):  # Got a tuple of jobs
+            self._project = jobs[0]._project
+        else:  # Got a project
+            self._project = jobs
+
+    def __hash__(self):
+        blob = str(self._project)
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+
+def get_aggregate_id(jobs):
+    """Generate hashed id for an aggregate of jobs
+
+    :param jobs:
+        The signac job handles
+    :type jobs:
+        tuple
+    """
+    if len(jobs) == 1:
+        return str(jobs[0])  # Return job id as it's already unique
```

Comment:
I would prefer to return `jobs[0].id`, which is strictly correct and doesn't depend on us not changing the string representation of a `Job`.